src/rt/util/typeinfo.d

```diff
@@ -1,107 +1,64 @@
 /**
- * This module contains utilities for TypeInfo implementation.
+ * A few predefined implementations for primitive types and arrays thereof. Also a couple of helpers.
  *
  * Copyright: Copyright Kenji Hara 2014-.
- * License:   <a href="http://www.boost.org/LICENSE_1_0.txt">Boost License 1.0</a>.
+ * License:   <a href="https://boost.org/LICENSE_1_0.txt">Boost License 1.0</a>.
  * Authors:   Kenji Hara
  */
 module rt.util.typeinfo;
 static import core.internal.hash;
 
-template Floating(T)
-if (is(T == float) || is(T == double) || is(T == real))
+// Three-way compare for integrals: negative if `lhs < rhs`, positive if `lhs > rhs`, 0 otherwise.
+pragma(inline, true)
+private int cmp3(T)(const T lhs, const T rhs)
+if (__traits(isIntegral, T))
 {
-  pure nothrow @safe:
-
-    bool equals(T f1, T f2)
-    {
-        return f1 == f2;
-    }
-
-    int compare(T d1, T d2)
-    {
-        if (d1 != d1 || d2 != d2) // if either are NaN
-        {
-            if (d1 != d1)
-            {
-                if (d2 != d2)
-                    return 0;
-                return -1;
-            }
-            return 1;
-        }
-        return (d1 == d2) ? 0 : ((d1 < d2) ? -1 : 1);
-    }
-
-    public alias hashOf = core.internal.hash.hashOf;
+    static if (T.sizeof < int.sizeof)
+        // Taking the difference will always fit in an int.
+        return int(lhs) - int(rhs);
+    else
+        return (lhs > rhs) - (lhs < rhs);
 }
-template Floating(T)
-if (is(T == cfloat) || is(T == cdouble) || is(T == creal))
-{
-  pure nothrow @safe:
-
-    bool equals(T f1, T f2)
-    {
-        return f1 == f2;
-    }
 
-    int compare(T f1, T f2)
-    {
-        int result;
-
-        if (f1.re < f2.re)
-            result = -1;
-        else if (f1.re > f2.re)
-            result = 1;
-        else if (f1.im < f2.im)
-            result = -1;
-        else if (f1.im > f2.im)
-            result = 1;
-        else
-            result = 0;
-        return result;
-    }
-
-    public alias hashOf = core.internal.hash.hashOf;
+// Three-way compare for real fp types. NaN is smaller than all valid numbers.
+// Code is small and fast, see https://godbolt.org/z/fzb877
+pragma(inline, true)
+private int cmp3(T)(const T d1, const T d2)
+if (is(T == float) || is(T == double) || is(T == real))
+{
+    if (d2 != d2)
```

Comment:
This codecov warning seems spurious - the unittest below exercise all paths in this function. cc @WalterBright